File: brd-android/app/src/main/java/com/breadwallet/mobius/QueuedConsumer.kt
/**
 * BreadWallet
 *
 * Created by Drew Carlson <drew.carlson@breadwallet.com> on 8/13/19.
 * Copyright (c) 2021 Breadwinner AG
 *
 * SPDX-License-Identifier: BUSL-1.1
 */
package com.breadwallet.mobius

import com.spotify.mobius.functions.Consumer
import kt.mobius.functions.Consumer as KtConsumer


/**
 * Collects events and passes them in order to a new consumer via [dequeueAll].
 */
class QueuedConsumer<V> : Consumer<V> {

    private val queue = arrayListOf<V>()

    override fun accept(value: V) = synchronized<Unit>(queue) {
        queue.add(value)
    }

    fun dequeueAll(target: Consumer<V>) = synchronized(queue) {
        queue.forEach(target::accept)
        queue.clear()
    }
}

class QueuedConsumerKt<V> : KtConsumer<V> {

    private val queue = arrayListOf<V>()

    override fun accept(value: V) = synchronized<Unit>(queue) {
        queue.add(value)
    }

    fun dequeueAll(target: KtConsumer<V>) = synchronized(queue) {
        queue.forEach(target::accept)
        queue.clear()
    }
}

File: brd-ios/breadwallet/Cosmos/Consumers/QueuedConsumer.swift
//
//  QueuedConsumer.swift
//  breadwallet
//
//  Created by blockexplorer <michael.inger@brd.com> on 2/26/21.
//  Copyright (c) 2021 Breadwinner AG
//
//  SPDX-License-Identifier: BUSL-1.1
//

import Foundation
import Cosmos

class QueuedConsumer: Consumer {

    var events: [Any?] = []

    func dequeueAll(consumer: Consumer) {
        let dequeued = self.events
        events = []
        dequeued.forEach { consumer.accept(value: $0) }
    }

    func accept(value: Any?) {
        events.append(value)
    }
}
